### DISCUSSION TITLE
[v0] How to translate godwoken_short_address to eth_address?

### DISCUSSION THREAD

**Discussion by Flouse** [nervosnetwork/godwoken#1069] [Q&A] [answered]
## Scenario Description
A developer wants to migrate an NFT contract from Godwoken v0 to Godwoken v1.
The contract recorded the addresses of users who minted tokens. 
But these addresses do not correspond to Ethereum addresses. 

By querying the contract in v0, the developer has a list of addresses (a.k.a godwoken_short_address, which is calculated from short_address = blake2b(script.as_slice())[0..20]).

Now the developer wants to find out the corresponding eth_address of each godwoken_short_address.
And then he can distribute tokens to the eth_addresses on Godwoken v1.

**Comment by Flouse** [comment, marked as answer]
# Solution

Let's say you have a godwoken_short_address `0xaa61ffa7c181bd3e23cb62c91291178e70c7b311`.

## 1. gw_get_script_hash_by_short_address
```bash
curl -X POST https://mainnet.godwoken.io/rpc \
  -H "Content-Type: application/json" \
  --data '{"jsonrpc":"2.0","method":"gw_get_script_hash_by_short_address","params": ["0xaa61ffa7c181bd3e23cb62c91291178e70c7b311"], "id":1}'
```

### Result
-> [0xaa61ffa7c181bd3e23cb62c91291178e70c7b311715511e299776a199f9f9c72](https://v0.gwscan.com/account/0xb67689858f88b58eeea9324a52a934b59f2beb22?search=0xaa61ffa7c181bd3e23cb62c91291178e70c7b311715511e299776a199f9f9c72)


## 2. gw_get_script
```bash
curl -X POST https://mainnet.godwoken.io/rpc \
  -H "Content-Type: application/json" \
  --data '{"jsonrpc":"2.0","method":"gw_get_script","params": ["0xaa61ffa7c181bd3e23cb62c91291178e70c7b311715511e299776a199f9f9c72"], "id":1}'

# Result
-> {"jsonrpc":"2.0","id":1,"result":{
"code_hash":"0x1563080d175bf8ddd44a48e850cecf0c0b4575835756eb5ffd53ad830931b9f9",
"hash_type":"type",
"args":"0x40d73f0d3c561fcaae330eabc030d8d96a9d0af36d0c5114883658a350cb9e3bb67689858f88b58eeea9324a52a934b59f2beb22"
}}
```

## 3. The eth_address of the account is the last 20 bytes of `result.args`:

0x40d73f0d3c561fcaae330eabc030d8d96a9d0af36d0c5114883658a350cb9e3bb67689858f88b58eeea9324a52a934b59f2beb22
=>
eth_address: [0xb67689858f88b58eeea9324a52a934b59f2beb22](https://v0.gwscan.com/account/0xb67689858f88b58eeea9324a52a934b59f2beb22)



## References
- [A short script for reference](https://github.com/godwokenrises/godwoken/discussions/1069#discussioncomment-5911669)
- https://github.com/godwokenrises/godwoken-polyjuice/tree/0.8-rc#address-used-in-polyjuice
- https://github.com/godwokenrises/godwoken-polyjuice/blob/0.8-rc/docs/EVM-compatible.md#different-address-type

**Reply by MaksHladun** [reply]
godwoken short address  0x650dfa9370E2179Ef31174b8fE4941A86146c14C

**Reply by MaksHladun** [reply]
Please tell me what could be the reason that the script generally worked fine, but some of the godwoken short addresses don't return eth addresses. 
The script returns an error when I pass one of these short addresses to it "Error: JSONRPCError: server error {"code":-32602,"message":"Invalid params"}"
and postman returns ""result": null"
What could be wrong is that these short addresses do not return the eth address?

here is one of godwoken short address "0x650dfa9370E2179Ef31174b8fE4941A86146c14C" that does not return an eth address.
Please help me with this

**Reply by jjyr** [reply]
Godwoken v0 converts the `eth address` and `short address` when the user deposits on layer-2.

The address `0x650dfa9370E2179Ef31174b8fE4941A86146c14C` never deposit on the Godwoken v0, so the endpoint returns null.  https://v0.gwscan.com/zh-CN/account/0x650dfa9370e2179ef31174b8fe4941a86146c14c?tab=transactions

**Reply by MaksHladun** [reply]
> Godwoken v0 converts the `eth address` and `short address` when the user deposits on layer-2.
> 
> The address `0x650dfa9370E2179Ef31174b8fE4941A86146c14C` never deposit on the Godwoken v0, so the endpoint returns null. https://v0.gwscan.com/zh-CN/account/0x650dfa9370e2179ef31174b8fe4941a86146c14c?tab=transactions

It turns out that it will not be possible to transform the eth address from this short address?

Because this address has nft tokens and now it is unclear how it minted them if the user never made a deposit on Godwoken v0.

**Reply by Flouse** [reply]
Basically, you're right.

Regarding the address 0x650dfa9370E2179Ef31174b8fE4941A86146c14C, it should be noted that its ownership is currently unknown, as no transactions have been sent from the address on Godwoken v0.

### Here's the advice we gave:

If there are only a few addresses with unknown ownership, it is recommended to skip migrating them for now. However, if a user can provide proof of ownership for this address on Godwoken v0 by a deposit on that account, you can then mint the corresponding NFT to that address.

**Reply by MaksHladun** [reply]
Ok. Thanks for help.

**Comment by Flouse** [comment]
## ~~Solution 2: use poly_getEthAddressByGodwokenShortAddress~~

```bash
curl -X POST https://mainnet.godwoken.io/rpc \
  -H "Content-Type: application/json" \
  --data '{"jsonrpc":"2.0","method":"poly_getEthAddressByGodwokenShortAddress","params": ["address_in_v0_contract"], "id":1}'
```

**Reply by MaksHladun** [reply]
Please tell me why the second solution may not work ( return "result": null)? The first works, but I would like to somehow facilitate the process when there are several hundreds of addresses.

**Reply by classicalliu** [reply]
A short script for reference 

```js
const { RPC } = require("ckb-js-toolkit")

async function main() {
  const rpc = new RPC("https://mainnet.godwoken.io/rpc")
  const shortAddress = process.argv[2];
  if (shortAddress == null) {
    throw new Error("please input short address")
  }
  const ethAddress = await shortAddressToEthAddress(shortAddress, rpc)
  console.log(`${shortAddress} --> ${ethAddress}`)
}
main()

async function shortAddressToEthAddress(shortAddress, rpc) {
  const scriptHash = await rpc.gw_get_script_hash_by_short_address(shortAddress)
  const script = await rpc.gw_get_script(scriptHash)
  const scriptArgs = script.args
  const ethAddress = "0x" + scriptArgs.slice(scriptArgs.length - 40, scriptArgs.length)
  return ethAddress
}
```

```console
➜ node short-address-to-eth-address.js 0xaa61ffa7c181bd3e23cb62c91291178e70c7b311
0xaa61ffa7c181bd3e23cb62c91291178e70c7b311 --> 0xb67689858f88b58eeea9324a52a934b59f2beb22
```

**Reply by MaksHladun** [reply]
Thank you very much. It works for me
